<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span><ul class="toc-item"><li><span><a href="#Оптимизация-данных" data-toc-modified-id="Оптимизация-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Оптимизация данных</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Пропуски-в-столбцах-days_employed-и-total_income" data-toc-modified-id="Пропуски-в-столбцах-days_employed-и-total_income-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Пропуски в столбцах days_employed и total_income</a></span></li><li><span><a href="#Аномалии-в-данных" data-toc-modified-id="Аномалии-в-данных-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Аномалии в данных</a></span></li><li><span><a href="#Замена-типа-данных-в-столбце-total_income" data-toc-modified-id="Замена-типа-данных-в-столбце-total_income-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Замена типа данных в столбце total_income</a></span></li><li><span><a href="#Дубликаты" data-toc-modified-id="Дубликаты-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Дубликаты</a></span></li><li><span><a href="#Создадим-два-новых-датафрейма" data-toc-modified-id="Создадим-два-новых-датафрейма-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Создадим два новых датафрейма</a></span></li><li><span><a href="#Функция-категории-дохода" data-toc-modified-id="Функция-категории-дохода-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Функция категории дохода</a></span></li><li><span><a href="#Функция-категоризации-целей-кредита" data-toc-modified-id="Функция-категоризации-целей-кредита-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Функция категоризации целей кредита</a></span></li></ul></li><li><span><a href="#Ответы-на-вопросы" data-toc-modified-id="Ответы-на-вопросы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Ответы на вопросы</a></span></li></ul></div>

# Исследование надёжности заёмщиков
# Описание проекта
**Заказчик** — кредитный отдел банка. 

**Входные данные от банка** — статистика о платёжеспособности клиентов.

**Цель** — разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок.

Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

**Дополнительный уровень:** 

Совершить оптимизацию данных для экономия памяти.

# Описание данных
* `children` — количество детей в семье
* `days_employed` — общий трудовой стаж в днях
* `dob_years` — возраст клиента в годах
* `education` — уровень образования клиента
* `education_id` — идентификатор уровня образования
* `family_status` — семейное положение
* `family_status_id` — идентификатор семейного положения
* `gender` — пол клиента
* `income_type` — тип занятости
* `debt` — имел ли задолженность по возврату кредитов
* `total_income` — ежемесячный доход
* `purpose` — цель получения кредита

## Обзор данных

### Оптимизация данных

In [1]:
# загрузим библиотеки
import pandas as pd
import numpy as np

In [2]:
# посмотрим на первые 5 строк
df = pd.read_csv('/datasets/data.csv')
df.head(5)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

In [3]:
# общая информация
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [4]:
# более подробные данные по используемой памяти
df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 11.3 MB


In [5]:
'''Исследуем использование памяти разными типами данных. 
Начнём со среднего показателя использования памяти по разным типам данных.'''
for dtype in ['float','int','object']:
    selected_dtype = df.select_dtypes(include=[dtype])
    mean_usage_b = selected_dtype.memory_usage(deep=True).mean()
    mean_usage_mb = mean_usage_b / 1024 ** 2
    print("Average memory usage for {} columns: {:03.2f} MB".format(dtype, mean_usage_mb))


Average memory usage for float columns: 0.11 MB
Average memory usage for int columns: 0.00 MB
Average memory usage for object columns: 1.69 MB


Эти сведения дают нам понять то, что большая часть памяти уходит на 5 столбцов, хранящих объектные значения.

In [6]:
'''Для проверки минимального и максимального значения, 
подходящего для хранения с использованием каждого целочисленного подтипа, 
можно воспользоваться методом numpy.iinfo(). Рассмотрим пример:'''
int_types = ["uint8", "int8", "int16"]
for it in int_types:
    print(np.iinfo(it))

Machine parameters for uint8
---------------------------------------------------------------
min = 0
max = 255
---------------------------------------------------------------

Machine parameters for int8
---------------------------------------------------------------
min = -128
max = 127
---------------------------------------------------------------

Machine parameters for int16
---------------------------------------------------------------
min = -32768
max = 32767
---------------------------------------------------------------



Тут можно обратить внимание на различие между типами uint (беззнаковое целое) и int (целое число со знаком). Оба типа имеют одинаковую ёмкость, но, при хранении в столбцах только положительных значений, беззнаковые типы позволяют эффективнее расходовать память.

In [7]:
# создадим функцию, которая поможет нам сэкономить немного времени на выяснении сколько памяти используется.

def mem_usage(pandas_obj):
    if isinstance(pandas_obj,pd.DataFrame):
        usage_b = pandas_obj.memory_usage(deep=True).sum()
    else: # исходим из предположения о том, что если это не DataFrame, то это Series
        usage_b = pandas_obj.memory_usage(deep=True)
    usage_mb = usage_b / 1024 ** 2 # преобразуем байты в мегабайты
    return "{:03.2f} MB".format(usage_mb)

In [8]:
gl_int = df.select_dtypes(include=['int'])
converted_int = gl_int.apply(pd.to_numeric,downcast='unsigned')

print(mem_usage(gl_int))
print(mem_usage(converted_int))

compare_ints = pd.concat([gl_int.dtypes,converted_int.dtypes],axis=1)
compare_ints.columns = ['before','after']
compare_ints.apply(pd.Series.value_counts)

0.00 MB
0.00 MB


Empty DataFrame
Columns: [before, after]
Index: []

In [9]:
gl_float = df.select_dtypes(include=['float'])
converted_float = gl_float.apply(pd.to_numeric,downcast='float')

print(mem_usage(gl_float))
print(mem_usage(converted_float))

compare_floats = pd.concat([gl_float.dtypes,converted_float.dtypes],axis=1)
compare_floats.columns = ['before','after']
compare_floats.apply(pd.Series.value_counts)

0.33 MB
0.16 MB


before  after
float32     NaN    2.0
float64     2.0    NaN

In [10]:
optimized_gl = df.copy()

optimized_gl[converted_int.columns] = converted_int
optimized_gl[converted_float.columns] = converted_float

print(mem_usage(df))
print(mem_usage(optimized_gl))

11.28 MB
11.12 MB


In [11]:
'''Для того чтобы понять, где именно мы сможем воспользоваться категориальными данными для снижения потребления памяти,
выясним количество уникальных значений в столбцах, хранящих значения объектных типов:'''
gl_obj = df.select_dtypes(include=['object']).copy()
gl_obj.describe()

education    family_status gender income_type  purpose
count      21525            21525  21525       21525    21525
unique        15                5      3           8       38
top      среднее  женат / замужем      F   сотрудник  свадьба
freq       13750            12380  14236       11119      797

In [12]:
dow = gl_obj.family_status
print(dow.head())

dow_cat = dow.astype('category')
print(dow_cat.head())

0     женат / замужем
1     женат / замужем
2     женат / замужем
3     женат / замужем
4    гражданский брак
Name: family_status, dtype: object
0     женат / замужем
1     женат / замужем
2     женат / замужем
3     женат / замужем
4    гражданский брак
Name: family_status, dtype: category
Categories (5, object): ['Не женат / не замужем', 'в разводе', 'вдовец / вдова', 'гражданский брак', 'женат / замужем']


In [13]:
dow_cat.head().cat.codes

0    4
1    4
2    4
3    4
4    3
dtype: int8

Тут можно заметить то, что каждому уникальному значению назначено целочисленное значение, и то, что столбец теперь имеет тип int8.

In [14]:
print(mem_usage(dow))
print(mem_usage(dow_cat))

2.93 MB
0.02 MB


Потребление памяти до и после преобразования столбца family_status к типу category.

In [15]:
'''Создадим цикл, который перебирает все столбцы, хранящие данные типа object, 
выясняет не превышает ли число уникальных значений в столбцах 50% 
и если это так, преобразует их в тип category.'''
converted_obj = pd.DataFrame()

for col in gl_obj.columns:
    num_unique_values = len(gl_obj[col].unique())
    num_total_values = len(gl_obj[col])
    if num_unique_values / num_total_values < 0.5:
        converted_obj.loc[:,col] = gl_obj[col].astype('category')
    else:
        converted_obj.loc[:,col] = gl_obj[col]

Сравним то, что получилось после оптимизации, с тем, что было раньше:

In [16]:
print(mem_usage(gl_obj))
print(mem_usage(converted_obj))

compare_obj = pd.concat([gl_obj.dtypes,converted_obj.dtypes],axis=1)
compare_obj.columns = ['before','after']
compare_obj.apply(pd.Series.value_counts)

12.34 MB
0.11 MB


before  after
object       5.0    NaN
category     NaN    1.0
category     NaN    1.0
category     NaN    1.0
category     NaN    1.0
category     NaN    1.0

In [17]:
# проанализируем то, на какой уровень использования памяти мы вышли, если сравнить то, что получилось
optimized_gl[converted_obj.columns] = converted_obj

print(mem_usage(optimized_gl))

1.10 MB


In [18]:
dtypes = optimized_gl.dtypes

dtypes_col = dtypes.index
dtypes_type = [i.name for i in dtypes.values]

column_types = dict(zip(dtypes_col, dtypes_type))

'''Возьмём пары ключ/значение из словаря
и аккуратно их выведем
'''
preview = first2pairs = {key:value for key,value in list(column_types.items())}
import pprint
pp = pp = pprint.PrettyPrinter(indent=4)
pp.pprint(preview)

{   'children': 'int64',
    'days_employed': 'float32',
    'debt': 'int64',
    'dob_years': 'int64',
    'education': 'category',
    'education_id': 'int64',
    'family_status': 'category',
    'family_status_id': 'int64',
    'gender': 'category',
    'income_type': 'category',
    'purpose': 'category',
    'total_income': 'float32'}


In [19]:
read_and_optimized = pd.read_csv('/datasets/data.csv',dtype=column_types,infer_datetime_format=True)

print(mem_usage(read_and_optimized))
display(read_and_optimized.head())
u_memory = 1 - (1.10 / 11.3)
print(f'Экономия памяти: {u_memory:.2%}')

1.10 MB


children  days_employed  dob_years education  education_id  \
0         1   -8437.672852         42    высшее             0   
1         1   -4024.803711         36   среднее             1   
2         0   -5623.422852         33   Среднее             1   
3         3   -4124.747070         32   среднее             1   
4         0  340266.062500         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.640625   
1   женат / замужем                 0      F   сотрудник     0  112080.015625   
2   женат / замужем                 0      M   сотрудник     0  145885.953125   
3   женат / замужем                 0      M   сотрудник     0  267628.562500   
4  гражданский брак                 1      F   пенсионер     0  158616.078125   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

Экономия памяти: 90.27%


In [20]:
read_and_optimized.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   children          21525 non-null  int64   
 1   days_employed     19351 non-null  float32 
 2   dob_years         21525 non-null  int64   
 3   education         21525 non-null  category
 4   education_id      21525 non-null  int64   
 5   family_status     21525 non-null  category
 6   family_status_id  21525 non-null  int64   
 7   gender            21525 non-null  category
 8   income_type       21525 non-null  category
 9   debt              21525 non-null  int64   
 10  total_income      19351 non-null  float32 
 11  purpose           21525 non-null  category
dtypes: category(5), float32(2), int64(5)
memory usage: 1.1 MB


In [21]:
#read_and_optimized.dtypes.values
#read_and_optimized.dtypes.value_counts()
#read_and_optimized.select_dtypes(include = ['float64', 'int64', 'float32', 'category'])
#level_map = {'F': 'женский', 'M': 'мужской', 'XNA': 'неизвестно'}
#read_and_optimized['c_gender'] = read_and_optimized['gender'].map(level_map)
#read_and_optimized

**1. Поиск пропусков:**

In [22]:
read_and_optimized.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**2. Поиск аномалий:**

In [23]:
# посмотрим на данные( 5 первых строк)
read_and_optimized[:5]

children  days_employed  dob_years education  education_id  \
0         1   -8437.672852         42    высшее             0   
1         1   -4024.803711         36   среднее             1   
2         0   -5623.422852         33   Среднее             1   
3         3   -4124.747070         32   среднее             1   
4         0  340266.062500         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.640625   
1   женат / замужем                 0      F   сотрудник     0  112080.015625   
2   женат / замужем                 0      M   сотрудник     0  145885.953125   
3   женат / замужем                 0      M   сотрудник     0  267628.562500   
4  гражданский брак                 1      F   пенсионер     0  158616.078125   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

Замечено отрицательное количество дней трудового стажа в столбце `days_employed`.

**3. Поиск дубликатов:**

In [24]:
print(f'Число явных дубликатов: {read_and_optimized.duplicated().sum()}')
read_and_optimized[read_and_optimized.duplicated()][:5]

Число явных дубликатов: 54


children  days_employed  dob_years education  education_id  \
2849         0            NaN         41   среднее             1   
4182         1            NaN         34    ВЫСШЕЕ             0   
4851         0            NaN         60   среднее             1   
5557         0            NaN         58   среднее             1   
7808         0            NaN         57   среднее             1   

         family_status  family_status_id gender income_type  debt  \
2849   женат / замужем                 0      F   сотрудник     0   
4182  гражданский брак                 1      F   сотрудник     0   
4851  гражданский брак                 1      F   пенсионер     0   
5557  гражданский брак                 1      F   пенсионер     0   
7808  гражданский брак                 1      F   пенсионер     0   

      total_income                  purpose  
2849           NaN  покупка жилья для семьи  
4182           NaN                  свадьба  
4851           NaN                  свадьба  
5557           NaN          сыграть свадьбу  
7808           NaN    на проведение свадьбы

**В столбце `education` есть одни и те же значения(неявные дубликаты), но записанные по-разному: с использованием заглавных и строчных букв.**

### Вывод

Выполнен ***дополнительный уровень*** для экономия памяти. Экономия памяти составила: 90.27%.

**Найденные проблемы в данных:**
1. В двух столбцах есть пропущенные значения. Один из них — `days_employed`, второй - `total_income`.
2. Замечена аномалия в данных, отрицательное количество дней трудового стажа в столбце `days_employed`. Так же есть  аномальное значение "-1" в  столбце `children`.
3. Столбец `education` содержит неявные дубликаты, значения повторяются, но записаны в разных регистрах. Так же в данных присутствует 54 явных дубликата.

## Предобработка данных

### Пропуски в столбцах days_employed и total_income

In [25]:
print(f'''Пропусков в столбце `days_employed` - {read_and_optimized.days_employed.isna().sum()}
Это {round(read_and_optimized.days_employed.isna().sum() / len(read_and_optimized) * 100, 2)}% от общего числа данных.''')

Пропусков в столбце `days_employed` - 2174
Это 10.1% от общего числа данных.


In [26]:
print(f'''Пропусков в столбце `total_income` до изменений - {read_and_optimized.total_income.isna().sum()}
Это {round(read_and_optimized.total_income.isna().sum() / len(read_and_optimized) * 100, 2)}% от общего числа данных.
Возможная причина появления пропусков в данных - человеческий фактор. 
Пропуски по обоим столбцам находятся в одинаковых строках.
Выглядят эти пропуски так:''')
read_and_optimized[read_and_optimized.total_income.isna()][:5]

Пропусков в столбце `total_income` до изменений - 2174
Это 10.1% от общего числа данных.
Возможная причина появления пропусков в данных - человеческий фактор. 
Пропуски по обоим столбцам находятся в одинаковых строках.
Выглядят эти пропуски так:


children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   

    total_income                           purpose  
12           NaN                   сыграть свадьбу  
26           NaN                       образование  
29           NaN  строительство жилой недвижимости  
41           NaN  сделка с подержанным автомобилем  
55           NaN                   сыграть свадьбу

In [27]:
print(f'''Строк до изменений: {read_and_optimized.shape[0]}
Столбцов до изменений: {read_and_optimized.shape[1]}''')
# заполним пропущенные значения в total_income медианным значением
read_and_optimized.total_income = read_and_optimized.total_income.fillna(read_and_optimized.total_income.median())

Строк до изменений: 21525
Столбцов до изменений: 12


In [28]:
print(f'''Строк после изменений: {read_and_optimized.shape[0]}
Столбцов после изменений: {read_and_optimized.shape[1]}
Пропусков в столбце `total_income` после замены - {read_and_optimized.total_income.isna().sum()}
Выглядит это так:''')
read_and_optimized[read_and_optimized.days_employed.isna()][:5]

Строк после изменений: 21525
Столбцов после изменений: 12
Пропусков в столбце `total_income` после замены - 0
Выглядит это так:


children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   

    total_income                           purpose  
12   145017.9375                   сыграть свадьбу  
26   145017.9375                       образование  
29   145017.9375  строительство жилой недвижимости  
41   145017.9375  сделка с подержанным автомобилем  
55   145017.9375                   сыграть свадьбу

**Заполнить пропуски медианным значением — лучшее решение для количественных переменных, потому что в таком случае можно перекрыть выбросы по значениям(как минимальным, так и максимальным) без весомых последствий для картины в целом.**

### Аномалии в данных

In [29]:
# до изменений
display(read_and_optimized[:5])
# исправим отрицательное количество дней трудового стажа в столбце days_employed и children
read_and_optimized.days_employed = abs(read_and_optimized.days_employed)
read_and_optimized.children = abs(read_and_optimized.children)
# заполним пропуски в days_employed медианным значением
read_and_optimized.days_employed = read_and_optimized.days_employed.fillna(read_and_optimized.days_employed.median())
# аномалия в данных по детям
read_and_optimized.children = read_and_optimized.children.replace(20, 2)
# неправильно указанный пол, исключим эту строку
read_and_optimized = read_and_optimized[read_and_optimized.gender != 'XNA']

children  days_employed  dob_years education  education_id  \
0         1   -8437.672852         42    высшее             0   
1         1   -4024.803711         36   среднее             1   
2         0   -5623.422852         33   Среднее             1   
3         3   -4124.747070         32   среднее             1   
4         0  340266.062500         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.640625   
1   женат / замужем                 0      F   сотрудник     0  112080.015625   
2   женат / замужем                 0      M   сотрудник     0  145885.953125   
3   женат / замужем                 0      M   сотрудник     0  267628.562500   
4  гражданский брак                 1      F   пенсионер     0  158616.078125   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

In [30]:
# после изменений
read_and_optimized[:5]

children  days_employed  dob_years education  education_id  \
0         1    8437.672852         42    высшее             0   
1         1    4024.803711         36   среднее             1   
2         0    5623.422852         33   Среднее             1   
3         3    4124.747070         32   среднее             1   
4         0  340266.062500         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.640625   
1   женат / замужем                 0      F   сотрудник     0  112080.015625   
2   женат / замужем                 0      M   сотрудник     0  145885.953125   
3   женат / замужем                 0      M   сотрудник     0  267628.562500   
4  гражданский брак                 1      F   пенсионер     0  158616.078125   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

### Замена типа данных в столбце total_income

In [31]:
# до изменений
read_and_optimized.info()
# замена типа данных на int
read_and_optimized.total_income = read_and_optimized.total_income.astype('int64')

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21524 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   children          21524 non-null  int64   
 1   days_employed     21524 non-null  float32 
 2   dob_years         21524 non-null  int64   
 3   education         21524 non-null  category
 4   education_id      21524 non-null  int64   
 5   family_status     21524 non-null  category
 6   family_status_id  21524 non-null  int64   
 7   gender            21524 non-null  category
 8   income_type       21524 non-null  category
 9   debt              21524 non-null  int64   
 10  total_income      21524 non-null  float32 
 11  purpose           21524 non-null  category
dtypes: category(5), float32(2), int64(5)
memory usage: 1.3 MB


In [32]:
# после изменений
read_and_optimized.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21524 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   children          21524 non-null  int64   
 1   days_employed     21524 non-null  float32 
 2   dob_years         21524 non-null  int64   
 3   education         21524 non-null  category
 4   education_id      21524 non-null  int64   
 5   family_status     21524 non-null  category
 6   family_status_id  21524 non-null  int64   
 7   gender            21524 non-null  category
 8   income_type       21524 non-null  category
 9   debt              21524 non-null  int64   
 10  total_income      21524 non-null  int64   
 11  purpose           21524 non-null  category
dtypes: category(5), float32(1), int64(6)
memory usage: 1.3 MB


### Дубликаты

**Явные дубликаты**

In [33]:
# до изменений
display(read_and_optimized.shape)
# удаление явных дубликатов
read_and_optimized = read_and_optimized.drop_duplicates().reset_index(drop=True)
# после изменений
display(read_and_optimized.shape)

(21524, 12)

(21470, 12)

**Неявные дубликаты**

In [34]:
# до изменений в education
list(read_and_optimized.education.unique())

['высшее',
 'среднее',
 'Среднее',
 'СРЕДНЕЕ',
 'ВЫСШЕЕ',
 'неоконченное высшее',
 'начальное',
 'Высшее',
 'НЕОКОНЧЕННОЕ ВЫСШЕЕ',
 'Неоконченное высшее',
 'НАЧАЛЬНОЕ',
 'Начальное',
 'Ученая степень',
 'УЧЕНАЯ СТЕПЕНЬ',
 'ученая степень']

In [35]:
# приведение к нижнему регистру education
read_and_optimized.education = read_and_optimized.education.str.lower().astype('category')
# после изменений в education
list(read_and_optimized.education.unique())

['высшее', 'среднее', 'неоконченное высшее', 'начальное', 'ученая степень']

In [36]:
# по итогу данные  выглядят так
read_and_optimized[5:]

children  days_employed  dob_years education  education_id  \
5             0     926.185852         27    высшее             0   
6             0    2879.202148         43    высшее             0   
7             0     152.779572         50   среднее             1   
8             2    6929.865234         35    высшее             0   
9             0    2188.756348         41   среднее             1   
...         ...            ...        ...       ...           ...   
21465         1    4529.316895         43   среднее             1   
21466         0  343937.406250         67   среднее             1   
21467         1    2113.346924         38   среднее             1   
21468         3    3112.481689         38   среднее             1   
21469         2    1984.507568         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
5      гражданский брак                 1      M   компаньон     0   
6       женат / замужем                 0      F   компаньон     0   
7       женат / замужем                 0      M   сотрудник     0   
8      гражданский брак                 1      F   сотрудник     0   
9       женат / замужем                 0      M   сотрудник     0   
...                 ...               ...    ...         ...   ...   
21465  гражданский брак                 1      F   компаньон     0   
21466   женат / замужем                 0      F   пенсионер     0   
21467  гражданский брак                 1      M   сотрудник     1   
21468   женат / замужем                 0      M   сотрудник     1   
21469   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose  
5            255763                 покупка жилья  
6            240525             операции с жильем  
7            135823                   образование  
8             95856         на проведение свадьбы  
9            144425       покупка жилья для семьи  
...             ...                           ...  
21465        224791             операции с жильем  
21466        155999          сделка с автомобилем  
21467         89672                  недвижимость  
21468        244093  на покупку своего автомобиля  
21469         82047         на покупку автомобиля  

[21465 rows x 12 columns]

Поиск дубликатов произодился в двух направлениях, по явным и неявным дубликатам. В этом помогли: метод `.duplicated().sum()` по явным дубликатам и метод `.unique()` для поиска неявных дубликатов.

Возможные причины появления дубликатов: человеческий фактор(неявные дубликаты) и технологический фактор(явные дубликаты).

In [37]:
read_and_optimized.education.unique()

['высшее', 'среднее', 'неоконченное высшее', 'начальное', 'ученая степень']
Categories (5, object): ['высшее', 'среднее', 'неоконченное высшее', 'начальное', 'ученая степень']

In [38]:
read_and_optimized.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21470 entries, 0 to 21469
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   children          21470 non-null  int64   
 1   days_employed     21470 non-null  float32 
 2   dob_years         21470 non-null  int64   
 3   education         21470 non-null  category
 4   education_id      21470 non-null  int64   
 5   family_status     21470 non-null  category
 6   family_status_id  21470 non-null  int64   
 7   gender            21470 non-null  category
 8   income_type       21470 non-null  category
 9   debt              21470 non-null  int64   
 10  total_income      21470 non-null  int64   
 11  purpose           21470 non-null  category
dtypes: category(5), float32(1), int64(6)
memory usage: 1.2 MB


### Создадим два новых датафрейма

In [39]:
list(read_and_optimized.purpose.unique())

['покупка жилья',
 'приобретение автомобиля',
 'дополнительное образование',
 'сыграть свадьбу',
 'операции с жильем',
 'образование',
 'на проведение свадьбы',
 'покупка жилья для семьи',
 'покупка недвижимости',
 'покупка коммерческой недвижимости',
 'покупка жилой недвижимости',
 'строительство собственной недвижимости',
 'недвижимость',
 'строительство недвижимости',
 'на покупку подержанного автомобиля',
 'на покупку своего автомобиля',
 'операции с коммерческой недвижимостью',
 'строительство жилой недвижимости',
 'жилье',
 'операции со своей недвижимостью',
 'автомобили',
 'заняться образованием',
 'сделка с подержанным автомобилем',
 'получение образования',
 'автомобиль',
 'свадьба',
 'получение дополнительного образования',
 'покупка своего жилья',
 'операции с недвижимостью',
 'получение высшего образования',
 'свой автомобиль',
 'сделка с автомобилем',
 'профильное образование',
 'высшее образование',
 'покупка жилья для сдачи',
 'на покупку автомобиля',
 'ремонт жилью',
 '

In [40]:
# каждому уникальному значению из education соответствует уникальное значение education_id
edu_dict = read_and_optimized[['education','education_id']]
edu_dict = edu_dict.drop_duplicates(keep='first').reset_index(drop=True).sort_values(by='education_id', ascending=True)
edu_dict

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

In [41]:
# каждому уникальному значению из family_status соответствует уникальное значение family_status_id — во втором.
fam_dict = read_and_optimized[['family_status','family_status_id']]
fam_dict = fam_dict.drop_duplicates(keep='first').reset_index(drop=True).sort_values(by='family_status_id', ascending=True)
fam_dict

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  Не женат / не замужем                 4

In [42]:
# удалим из исходного датафрейма столбцы education и family_status, оставив только их идентификаторы: 
del read_and_optimized['education']
del read_and_optimized['family_status']

read_and_optimized.sample(5)

children  days_employed  dob_years  education_id  family_status_id  \
12098         1    4379.530273         42             1                 1   
3986          0  388259.843750         59             1                 0   
12396         1    1783.316772         22             1                 0   
2089          0    2194.220459         64             1                 0   
1383          0  353802.812500         37             1                 2   

      gender income_type  debt  total_income                     purpose  
12098      F   компаньон     0        163330       заняться образованием  
3986       F   пенсионер     0        257994        покупка недвижимости  
12396      M   компаньон     0         68512           операции с жильем  
2089       F   сотрудник     0        145017    операции с недвижимостью  
1383       F   пенсионер     0        216452  строительство недвижимости

### Функция категории дохода

In [43]:
def total_income_category(total_income):
    """
    Функция возвращает категорию дохода(E,D,C,B,A) по значению total_income, используя правила:
    - 0–30000 — 'E';
    - 30001–50000 — 'D';
    - 50001–200000 — 'C';
    - 200001–1000000 — 'B';
    - 1000001 и выше — 'A'.
    Если что-то пошло не по плану вернёт 'Ошибка'
    """
    if 30000 >= total_income:
        return 'E'
    elif 30000 < total_income <= 50000:
        return 'D'
    elif 50000 < total_income <= 200000:
        return 'C'
    elif 200000 < total_income <= 1000000:
        return 'B'
    elif 1000000 < total_income:
        return 'A'
    else:
        return 'Ошибка'

In [44]:
# создадим новый столбец и применим функцию выше
read_and_optimized['total_income_category'] = read_and_optimized['total_income'].apply(total_income_category)

In [45]:
# результат
read_and_optimized.sample(5)

children  days_employed  dob_years  education_id  family_status_id  \
12187         0    5292.090332         34             1                 1   
8840          1    2194.220459         30             1                 0   
4275          1    2194.220459         34             1                 0   
9605          0     568.978455         52             1                 1   
7120          0  398099.406250         67             1                 0   

      gender income_type  debt  total_income                     purpose  \
12187      M   компаньон     0         81309             сыграть свадьбу   
8840       M   сотрудник     0        145017                       жилье   
4275       M   компаньон     0        145017  строительство недвижимости   
9605       M   сотрудник     0        179249        покупка недвижимости   
7120       M   пенсионер     0         60039        сделка с автомобилем   

      total_income_category  
12187                     C  
8840                      C  
4275                      C  
9605                      C  
7120                      C

### Функция категоризации целей кредита

In [46]:
def purpose_category(purpose):
    '''
    Функция, которая на основании данных из столбца purpose сформирует новый столбец purpose_category,
    в который войдут следующие категории:
    - 'операции с автомобилем',
    - 'операции с недвижимостью',
    - 'проведение свадьбы',
    - 'получение образования'.
    '''
    if 'автомобил' in purpose:
        return 'операции с автомобилем'
    elif 'недвижимост' in purpose:
        return 'операции с недвижимостью'
    elif 'жиль' in purpose:
        return 'операции с недвижимостью'
    elif 'свадьб' in purpose:
        return 'проведение свадьбы'
    elif 'образован' in purpose:
        return 'получение образования'
    else:
        return 'уникальный случай'

In [47]:
# применим функцию
read_and_optimized['purpose_category'] = read_and_optimized['purpose'].apply(purpose_category)
# результат 
read_and_optimized.sample(5)

children  days_employed  dob_years  education_id  family_status_id  \
7618          0    2297.167480         30             1                 4   
21172         0  347401.375000         60             1                 0   
10151         0  388462.687500         55             0                 0   
3813          1    3045.290527         26             1                 1   
171           0    9566.241211         60             1                 1   

      gender  income_type  debt  total_income                     purpose  \
7618       F    компаньон     0        120981       на покупку автомобиля   
21172      F    пенсионер     0        107546                  автомобили   
10151      F    пенсионер     0        126296  строительство недвижимости   
3813       F  госслужащий     0        131892       на проведение свадьбы   
171        F    компаньон     0        225197       на проведение свадьбы   

      total_income_category          purpose_category  
7618                      C    операции с автомобилем  
21172                     C    операции с автомобилем  
10151                     C  операции с недвижимостью  
3813                      C        проведение свадьбы  
171                       B        проведение свадьбы

## Ответы на вопросы

**1. Есть ли зависимость между количеством детей и возвратом кредита в срок?**

In [66]:
child_debt = read_and_optimized.pivot_table(index=['children'], values='debt', aggfunc='mean').reset_index()
child_debt['debt'] =  round(child_debt['debt'], 4) * 100
child_debt

children  debt
0         0  7.54
1         1  9.16
2         2  9.49
3         3  8.18
4         4  9.76
5         5  0.00

**2. Есть ли зависимость между семейным положением и возвратом кредита в срок?** 

In [71]:
fam_debt = fam_dict.merge(read_and_optimized)
fam_debt = fam_debt[['family_status', 'family_status_id', 'debt']]
#fam_debt.groupby(['family_status', 'debt'])['debt'].count()
fam_debt = fam_debt.pivot_table(index=['family_status'], values='debt', aggfunc='mean').reset_index()
fam_debt['debt'] =  round(fam_debt['debt'], 4) * 100
fam_debt

family_status  debt
0  Не женат / не замужем  9.75
1              в разводе  7.11
2         вдовец / вдова  6.57
3       гражданский брак  9.32
4        женат / замужем  7.54

**3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**

In [74]:
income_category_debt = read_and_optimized.pivot_table(index=['total_income_category'], values='debt', aggfunc='mean').reset_index()
income_category_debt['debt'] =  round(income_category_debt['debt'], 4) * 100
income_category_debt

total_income_category  debt
0                     A  8.00
1                     B  7.06
2                     C  8.48
3                     D  6.00
4                     E  9.09

**4. Как разные цели кредита влияют на его возврат в срок?**

In [79]:
purpose_category_debt = read_and_optimized.pivot_table(index=['purpose_category'], values='debt', aggfunc='mean').reset_index()
purpose_category_debt['debt'] =  round(purpose_category_debt['debt'], 4) * 100
purpose_category_debt

purpose_category  debt
0    операции с автомобилем  9.35
1  операции с недвижимостью  7.23
2     получение образования  9.22
3        проведение свадьбы  7.97